packages/server/src/script/fetchMGErrors.ts

```diff
@@ -0,0 +1,129 @@
+import { processEnvOrThrow } from '../common'
+import { MailgunHookBody, MailgunEventReason } from '../service/webhooks'
+import { promises as fs } from 'fs'
+
+interface EntryResponse {
+  data: MailgunHookBody
+}
+
+/**
+ * * The first is the starting date (inclusive) of the script;
+ * * Likewise, the second can be the end date (inclusive), or null if we
+ * want to fetch all records from the starting point until the present moment;
+ * * The third is a list of MGEventReason to be ignored while fetching logs;
+ */
+export type MGFetchErrorsArgs = [Date, Date | null, MailgunEventReason[]]
+// Only used to ease importing through require()
+interface MGFetchErrorsArgsExport { args: MGFetchErrorsArgs }
+
+const { args } = process.env.CI
+  ? require('./data/mgConditions.sample.ts') as MGFetchErrorsArgsExport
+  : require('./data/mgConditions.nogit.ts') as MGFetchErrorsArgsExport
+
+const projectId = processEnvOrThrow('GCLOUD_PROJECT')
+
+const fetchMGErrors = async (
+  fromIsoStamp: string,
+  untilIsoStamp: string | null,
+  ignoredEventReasons: MailgunEventReason[] = [],
+) => {
+  const logging = new Logging({ projectId })
+
+  const filter =
+`logName="projects/${projectId}/logs/mailgun-log"
+severity=(WARNING OR ERROR OR CRITICAL OR ALERT OR EMERGENCY)
+
+timestamp >= "${fromIsoStamp}"
+${untilIsoStamp ? `timestamp <= ${untilIsoStamp}` : ''}`
+
+  console.log('Querying all MG errors in our logs during the period, this might take a while...\n')
+  // The response comes in a nested array, flattening it to ease iterations
+  const entries = (await logging.log('mailgun-log').getEntries({ filter })).flat() as EntryResponse[]
+
+  const results: Record<string, string[]> = {}
+
+  entries.forEach(({ data }) => {
+    // Do not resend since the email doesn't exist, apparently the message
+    // for this event vary depending on the destination
+    //
+    // We don't use the delivery-status code here since email providers'
+    // implementation differ slightly from each other, for example at Google
+    // Gmail we have 550 for non existent emails, but on Yahoo the same error
+    // is represented by 554.
+    const message = data['event-data']?.['delivery-status']?.message
+
+    // Message can only be undefined/null in staging/dev environments, this
+    // happens when we've used MG's webhook testing hooks before (they don't
+    // contain delivery-statuses)
+    if (!message) {
+      return
+    }
+
+    // Found on production logs
+    if (message.indexOf('No SMTP service') >= 0) {
+      return
+    }
+    // At https://documentation.mailgun.com/en/latest/api-events.html?highlight=user%20variables#event-structure
+    // It seems that this is also the default message from Google's Gmail
+    if (message.indexOf('5.1.1 The email account that you tried to reach does not exist.') >= 0) {
+      return
+    }
+    // Used by Yahoo
+    if (message.indexOf('delivery error: dd Not a valid recipient') >= 0) {
```

Comment:
It seems crazy to me that Yahoo/Gmail would return some seemingly arbitrary and subject to change string with the failure reason, rather than some code which we'd expect to stay consistent.  See if you can figure out something more reliable than this.  That being said, it seems unlikely that this will change over the next few weeks ;)

Reply:
It seems unlikely for me to find a better alternative than relying on status messages to detect which emails exists/don't exist. 

From what I've seen Gmail hasn't changed its status message since a long time (the documentation on MG for instance has an example with the same message we're using right now, and you can find old links around the web with it, e.g. https://answers.microsoft.com/en-us/outlook_com/forum/oemail-osend/unable-to-send-email-message-and-received-a/f422ce38-1df7-43c5-962f-d7f51762f9c7)

Reply:
Ok, makes sense, just thought I'd suggest it -- but this seems like the best way.